packages/java/src/node/java-contribution.ts

```diff
@@ -70,9 +74,17 @@ export class JavaContribution extends BaseLanguageServerContribution {
         if (jarPaths.length === 0) {
             throw new Error('The Java server launcher is not found.');
         }
-
         const jarPath = path.resolve(serverPath, jarPaths[0]);
-        const workspacePath = path.resolve(os.tmpdir(), '_ws_' + new Date().getTime());
+
+        const uri = await this.workspaceService.getMostRecentlyUsedWorkspace();
```

Comment:
This is only bay luck returning the expected root

Reply:
I guess if we're reusing, somewhere in user home would be appropriate?

Reply:
Does Theia already have some way to get current workspace's name or URI?

Reply:
> Does Theia already have some way to get current workspace's name or URI?

Yes, however each frontend application is the one aware of what its own current workspace is - the backend doesn't know or keep track of this. This way, there can be multiple frontends, each with its own independently-set workspace (it's permitted to re-use a WS between clients), using the same stateless backend. 

One way where jdt.ls may differ from other LS we use is that it probably does not permit sharing a workspace between two or more clients. So for that case, we might still need to have different jdt.ls workspaces. However when a given client is reloaded, it would make sense for it to re-use the same jdt.ls workspace as before.

Reply:
@marcdumais-work To re-use the same jdt.ls workspace after reloading, backend has to know some identifier of the clients to create jdt.ls workspace. I thought about workspace URI as an identifier of frontend  application. But the problem here is that I know nothing about frontend app when jdt.ls is starting.

Reply:
> > Does Theia already have some way to get current workspace's name or URI?
> 
> Yes, however each frontend application is the one aware of what its own current workspace is - the backend doesn't know or keep track of this. This way, there can be multiple frontends, each with its own independently-set workspace (it's permitted to re-use a WS between clients), using the same stateless backend.

I think the stateless backend is kind not going to work: if we want to eventually share language server instances (which is a requirements if we want to scale to large workspaces), we will have to implement shared buffers in some way, so that backend will be stateful.

Reply:
> @marcdumais-work To re-use the same jdt.ls workspace after reloading, backend has to know some identifier of the clients to create jdt.ls workspace. I thought about workspace URI as an identifier of frontend application. But the problem here is that I know nothing about frontend app when jdt.ls is starting.

This scenario may be the one case where it's ok to rely on the first "recentRoot" entry in ```~/.theia/recentworkspace.json```. Or at least it's how we do it ATM - this is how Theia is able to keep the same workspace, after a frontend reload (e.g. open a few Theia clients each with a different WS, then cycle through the browser tabs and reload each one - they should each restart with their previously set workspace). The reason is that just before reloading, the terminating F-E app will save it's current workspace in that file, and the newly started app will immediately re-use that value, to set its workspace to the same thing. 

So I guess we are not totally stateless in the backend, w-r to workspace... But the workspace server is a singleton, with the bulk of the state being kept in the workspace service, in the F-E.

Reply:
>  if we want to eventually share language server instances

Interesting. Would a single jdt.ls instance be able to simultaneously serve two or more LS clients, for the same workspace?

Reply:
No, it would not: you can't initialize a language server more than once.

Reply:
> One way where jdt.ls may differ from other LS we use is that it probably does not permit sharing a workspace between two or more clients. 
Actually, I would expect this to be the case for many non-trivial languages servers: there would a some metadata stored and cached by the language server (think symbol index), and if the language server derives from some IDE tool (as jdt.ls), it probably assumes that it's working alone on that metadata.

Reply:
> No, it would not: you can't send the initialize message twice, for example. Also, there is no provision for opening multiple communication channels (socket or otherwise). Such multiplexing has to happen outside of the language server.

One could start one language server process that accepts many connections and then the language server internally could do all kinds of sharing and caching.

Reply:
> One could start one language server process that accepts many connections and then the language server internally could do all kinds of sharing and caching.
Yes, one could, but currently, language servers are not written that way.